File: src/main/kotlin/ru/proshik/english/quizlet/telegramBot/service/operation/Operation.java
package ru.proshik.english.quizlet.telegramBot.service.operation;

public interface Operation<T extends OperationResult, R extends OperationResultFormatter> {

    T init();

    OperationResult<?, ?> nextStep(String text, String data);

    // TODO rename builder on formatter
    R getFormatter();
}

File: src/main/kotlin/ru/proshik/english/quizlet/telegramBot/service/operation/Operation.kt
package ru.proshik.english.quizlet.telegramBot.service.operation

import org.telegram.telegrambots.meta.api.methods.BotApiMethod
import ru.proshik.english.quizlet.telegramBot.model.Account
import ru.proshik.english.quizlet.telegramBot.service.vo.OperationData
import java.io.Serializable

interface Operation {

    fun init(chatId: Long, account: Account): BotApiMethod<out Serializable>

    fun execute(chatId: Long,
                messageId: Int,
                callData: String,
                value: OperationData,
                account: Account): BotApiMethod<out Serializable>

}

